lib/charms/layer/aws.py

```diff
@@ -136,36 +136,64 @@ def enable_s3_write(application_name, instance_id, region):
 
 def cleanup(current_applications):
     log('Looking for unused AWS role and instance-profiles to cleanup')
-    cached_roles = _get_roles()
-    prefix = 'charm-aws-{}-'.format(os.environ['JUJU_MODEL_UUID'])
-    query = 'Roles[?starts_with(RoleName, `{}`)].RoleName'.format(prefix)
-    role_names = _aws('iam', 'list-roles', '--query', query)
+    model_uuid = os.environ['JUJU_MODEL_UUID']
+    prefix = 'charm-aws-{}-'.format(model_uuid)
+    role_names = _list_roles(model_uuid)
+    instance_profile_names = _list_instance_profiles(model_uuid)
     for role_name in role_names:
         application_name = role_name[len(prefix):]
         if application_name not in current_applications:
             log('Found: {}', role_name)
             _cleanup_role(role_name)
-            cached_roles.pop(role_name, None)
-    _update_roles()
+    for instance_profile_name in instance_profile_names:
+        application_name = instance_profile_name[len(prefix):]
+        if application_name not in current_applications:
+            log('Found: {}', instance_profile_name)
+            _cleanup_instance_profile(instance_profile_name)
 
 
 # Internal helpers
 
 
 class AWSError(Exception):
-    def __init__(self, message):
-        self.error_type = None
-        self.message = message
+    @classmethod
+    def get(cls, message):
+        error_type = None
         match = re.match(r'An error occurred \(([^)]+)\)', message)
         if match:
-            self.error_type = match.group(1)
+            error_type = match.group(1)
+        for error_cls in (DoesNotExistAWSError, AlreadyExistsAWSError):
+            if error_type in error_cls.error_types:
+                return error_cls(error_type, message)
+        return AWSError(error_type, message)
+
+    def __init__(self, error_type, message):
+        self.error_type = error_type
+        self.message = message
         super().__init__(message)
 
     def __str__(self):
         return self.message
 
 
-def _aws(cmd, subcmd, *args, ignore_errors=False):
+class DoesNotExistAWSError(AWSError):
+    # meta-error representing something not existing
+    error_types = [
```

Comment:
I believe in you.  I don't know the aws cli exceptions, but trust that you do ;)

Reply:
I couldn't find a reference, and there a likely other cases where some of these errors are raised, but these capture the cases I've encountered in testing.